src/sqlfluff/rules/L025.py

```diff
@@ -38,50 +59,78 @@ class Rule_L025(Rule_L020):
+
+            alias: AliasInfo
+            for alias in query.aliases:
+                if alias.aliased and alias.ref_str not in query.tbl_refs:
+                    # Unused alias. Report and fix.
+                    violations.append(self._report_unused_alias(alias))
+        return violations or None
 
-        alias: AliasInfo
-        for alias in table_aliases:
-            if alias.aliased and alias.ref_str not in tbl_refs:
-                fixes = [LintFix.delete(alias.alias_expression)]
-                # Walk back to remove indents/whitespaces
-                to_delete = (
-                    Segments(*alias.from_expression_element.segments)
-                    .reversed()
-                    .select(
-                        start_seg=alias.alias_expression,
-                        # Stop once we reach an other, "regular" segment.
-                        loop_while=sp.or_(sp.is_whitespace(), sp.is_meta()),
-                    )
-                )
-                fixes += [LintFix.delete(seg) for seg in to_delete]
-                violation_buff.append(
-                    LintResult(
-                        anchor=alias.segment,
-                        description="Alias {!r} is never used in SELECT statement.".format(
-                            alias.ref_str
-                        ),
-                        fixes=fixes,
-                    )
-                )
-        return violation_buff or None
+    @classmethod
+    def _analyze_table_aliases(cls, query: L025Query, dialect: Dialect):
+        # Get table aliases defined in query.
+        for selectable in query.selectables:
+            select_info = get_select_statement_info(selectable.selectable, dialect)
+            if select_info:
+                # Record the aliases.
+                query.aliases += select_info.table_aliases
+
+                # Look at each table reference; if its an alias reference,
+                # resolve the alias: could be an alias defined in "query"
+                # itself or an "ancestor" query.
+                for r in select_info.reference_buffer:
+                    for tr in r.extract_possible_references(level=r.ObjectReferenceLevel.TABLE):  # type: ignore
+                        # This function walks up the query's parent stack if necessary.
+                        # Reuses the existing "for r in references:" loop from L025.
+                        cls._resolve_and_mark_reference(query, tr.part)
+
+        # Visit children.
+        for child in query.children:
+            cls._analyze_table_aliases(cast(L025Query, child), dialect)
+
+    @classmethod
+    def _resolve_and_mark_reference(cls, query: L025Query, ref: str):
+        # Does this query define the referenced alias?
+        if any(ref == a.ref_str for a in query.aliases):
+            # Yes. Record the reference.
+            query.tbl_refs.add(ref)
+        elif query.parent:
+            # No. Recursively check the query's parent hierarchy.
```

Comment:
This is the key area of the fix: Now we'll correctly resolve a reference to a table alias that was defined in an "ancestor" query.